src/main/host/syscall/futex.c

```diff
@@ -0,0 +1,164 @@
+
+    // Futexes are 32 bits on all platforms
+    void* faddr = (uint32_t*)uaddrptr.val;
+    uint32_t expectedVal = (uint32_t)val;
+
+    // futex addr cannot be NULL
+    if (!faddr) {
+        debug("Futex addr cannot be NULL");
+        return (SysCallReturn){.state = SYSCALL_DONE, .retval.as_i64 = -EFAULT};
+    }
+
+    int result = 0;
+
+    switch (operation) {
+        case FUTEX_WAIT: {
+            debug("Handling FUTEX_WAIT operation %i", operation);
+
+            // This is a new wait operation on the futex for this thread.
+            // Check if a timeout was given in the syscall args.
+            const struct timespec* timeout;
+            if (timeoutptr.val) {
+                timeout =
+                    process_getReadablePtr(sys->process, sys->thread, timeoutptr, sizeof(*timeout));
+                // Bounds checking
+                if (!(timeout->tv_nsec >= 0 && timeout->tv_nsec <= 999999999)) {
+                    debug("A futex timeout was given, but the nanos value is out of range");
+                    result = -EINVAL;
+                    break;
+                }
+            } else {
+                timeout = NULL;
+            }
+
+            // TODO: does this load/compare need to be done atomically?
+            // `man 2 futex`: blocking via a futex is an atomic compare-and-block operation
+            const uint32_t* futexVal =
+                process_getReadablePtr(sys->process, sys->thread, uaddrptr, sizeof(uint32_t));
+
+            debug("Futex value is %" PRIu32 ", expected value is %" PRIu32, *futexVal, expectedVal);
+            if (*futexVal != expectedVal) {
+                debug("Futex values don't match, try again later");
+                result = -EAGAIN;
+                break;
+            }
+
+            // Check if we already have a futex
+            FutexTable* ftable = host_getFutexTable(sys->host);
+            Futex* futex = futextable_get(ftable, faddr);
+
+            if (_syscallhandler_wasBlocked(sys)) {
+                utility_assert(futex != NULL);
+
+                // We already blocked on wait, so this is either a timeout or wakeup
+                if (timeout && _syscallhandler_didListenTimeoutExpire(sys)) {
+                    // Timeout while waiting for a wakeup
+                    debug("Futex %p timeout out while waiting", faddr);
+                    result = -ETIMEDOUT;
+                } else {
+                    // Proper wakeup from another thread
+                    debug("Futex %p has been woke up", faddr);
+                    result = 0;
+                }
+
+                // Dynamically clean up the futex if needed
+                if (futex_getListenerCount(futex) == 0) {
+                    debug("Dynamically freed a futex object for futex addr %p", faddr);
+                    bool success = futextable_remove(ftable, futex);
+                    utility_assert(success);
+                }
+                break;
+            }
```

Comment:
It looks like when we intend to wake up a thread, we still fall through and end up returning SYSCALL_BLOCK? (and possibly using a now-freed futex pointer?)

It looks like there should be a `return` here?

Reply:
The break should cause us to break out of the switch statement, and then we return done immediately after that. (I'm going to add `// return SYSCALL_DONE` after the breaks to make that more clear.)

Reply:
Ah, I must have missed the `break` or incorrectly matched it up with the next brace.

Reply:
Optional - maybe it'd be worth breaking the (implemented) cases out into helper functions to make it easier to see the overall structure?